teos/gatekeeper.py

```diff
@@ -189,46 +228,117 @@ def add_update_appointment(self, user_id, uuid, ext_appointment):
+        """
+
+        with self.lock:
+            # Try to get the data from the cache
+            outdated_users = self.outdated_users_cache.get(block_height)
 
-        self.lock.release()
-        return self.registered_users.get(user_id).available_slots
+            # Get the data from registered_users otherwise
+            if not outdated_users:
+                outdated_users = {
+                    user_id: list(user_info.appointments.keys())
+                    for user_id, user_info in self.registered_users.items()
+                    if block_height == user_info.subscription_expiry + self.expiry_delta
+                }
 
-    def get_expired_appointments(self, block_height):
+            return outdated_users
+
+    def get_outdated_user_ids(self, block_height):
         """
-        Gets a list of appointments that expire at a given block height.
+        Returns a list of all user ids outdated at a given ``block_height``.
 
         Args:
-            block_height: the block height that wants to be checked.
+            block_height (:obj:`int`): the block height that wants to be checked.
 
         Returns:
-            :obj:`list`: A list of appointment uuids that will expire at ``block_height``.
+            :obj:`list`: A list of user ids whose subscription is outdated at ``block_height``.
+        """
+
+        return list(self.get_outdated_users(block_height).keys())
+
+    def get_outdated_appointments(self, block_height):
+        """
+        Returns a flattened list of all appointments outdated at a given ``block_height``, indistinguishably of their
+        user.
+
+        Args:
+            block_height (:obj:`int`): the block height that wants to be checked.
+
+        Returns:
+            :obj:`list`: A list of appointments whose subscription is outdated at ``block_height``.
+        """
+
+        return [
+            appointment_uuid
+            for user_appointments in self.get_outdated_users(block_height).values()
+            for appointment_uuid in user_appointments
+        ]
+
+    def update_outdated_users_cache(self, block_height):
+        """
+        Adds an entry corresponding to ``block_height`` to ``outdated_users_cache`` if the entry is missing, and removes
+        the oldest entry if the cache is full afterwards.
+
+        Args:
+            block_height (:obj:`int`): the block that acts as id for the new entry in the cache.
         """
 
-        expired_appointments = []
-        # Avoiding dictionary changed size during iteration
-        for user_id in list(self.registered_users.keys()):
-            if block_height == self.registered_users[user_id].subscription_expiry + self.expiry_delta:
-                expired_appointments.extend(self.registered_users[user_id].appointments)
+        outdated_users = self.get_outdated_users(block_height)
 
-        return expired_appointments
+        if block_height not in self.outdated_users_cache:
+            with self.lock:
```

Comment:
I would put all the code inside the lock, otherwise it could in principle happen that the code is stopped before getting the lock, meanwhile a reorg happens and a different block is processed, and the execution continues, while the state has changed. This would probably not be a problem here because the height after the re-org is likely the same, therefore the list of outdated users shouldn't change, but I think it's better to put the lock on the whole code and not have to think about it.

Reply:
Added